server/src/main/kotlin/core/action/bettinground/BettingAction.kt

```diff
@@ -0,0 +1,55 @@
+package core.action.bettinground
+import core.gameflow.HandState
+
+abstract class BettingAction(val type: ActionType) {
+
+    /**
+     * This method must not shift the activePlayer pointer as this logic
+     * is extracted to shiftActivePlayer method that is always called after innerApply.
+     */
+    protected abstract fun innerApply(handState: HandState): HandState
+    protected abstract fun innerIsLegal(handState: HandState): Boolean
+
+    fun apply(handState: HandState): HandState {
+        assert(isLegal(handState))
+
+        val newState = innerApply(handState)
+        return shiftActivePlayer(newState)
+    }
+
+    fun isLegal(handState: HandState): Boolean = (handState.activePlayer != null) and innerIsLegal(handState)
+
+    private fun shiftActivePlayer(handState: HandState): HandState {
+
+        handState.activePlayer!!
+        val nextDecisivePlayer = handState.nextDecisivePlayer(handState.activePlayer)
+
+        return when {
+
+            /* End action when there are no more decisive players. */
+            nextDecisivePlayer == null -> handState.copy(activePlayer = null)
+
+            /* End action when only one player is left in the hand. */
+            handState.playersInGame.size == 1 -> handState.copy(activePlayer = null)
+
+            /* End action when current action taker is the last decisive player in the hand. */
+            nextDecisivePlayer == handState.activePlayer -> handState.copy(activePlayer = null)
+
+            /* Proceed when next decisive player didn't have a chance to take any action yet. */
```

Comment:
we could change the order of proceed/end cases so that they are grouped together

Reply:
also ocd driven developement

Reply:
Nice as it would be, it's quite impossible, because the cases are not mutually exclusive and some of them need to be checked before others. Adding more conditions just to make sure they cover separate scenarios would be an overkill not worth the effort and would actually make the code less readable imho.

Reply:
OK